File: Tests/executors/CSharp/DatabaseExecutors/CRUDExecutor.cs
using Framework;
using Framework.Databases;
using Framework.Utility;

struct TableData : IDTO<TableData>
{
	public static TableData Create(IDictionary<string, SqlValue> row)
	{
		TableData result = new()
		{
			id = row["id"].GetValue<int>(),
			name = row["name"].GetValue<string>(),
			amount = row["amount"].GetValue<int>()
		};

		return result;
	}

	public int id;
	public string name;
	public int amount;
}

public class CRUDExecutor : StatelessExecutor
{
	public override void Init(ExecutorSettings settings)
	{
		settings.GetOrCreateDatabase("test_database").GetOrCreateTable
		(
			"test_table",
			"CREATE TABLE IF NOT EXISTS test_table (" +
			"id INTEGER PRIMARY KEY AUTOINCREMENT, " +
			"name VARCHAR(255) NOT NULL, " +
			"amount INTEGER NOT NULL)"
		);
	}

	public override void DoGet(HttpRequest request, HttpResponse response)
	{
		response.SetBody
		(
			new
			{
				data = request.GetTable("test_database", "test_table").ExecuteQuery<TableData>("SELECT * FROM test_table WHERE name = ?", WebFrameworkUtility.MakeSqlValues("glue"))
			}
		);
	}

	public override void DoPut(HttpRequest request, HttpResponse response)
	{
		Database database = request.GetDatabase("test_database");
		Table table = database.GetTable("test_table");
		Random random = new((int)DateTime.Now.Ticks);

		table.ExecuteQuery
		(
			"INSERT INTO test_table (name, amount) VALUES(?, ?)",
			WebFrameworkUtility.MakeSqlValues("glue", -1)
		);

		for (int i = 0; i < 10; i++)
		{
			table.ExecuteQuery
			(
				"INSERT INTO test_table (name, amount) VALUES(?, ?)",
				WebFrameworkUtility.MakeSqlValues("glue", random.Next() % 200)
			);
		}
	}

	public override void DoPatch(HttpRequest request, HttpResponse response)
	{
		Database database = request.GetDatabase("test_database");
		Table table = database.GetTable("test_table");

		table.ExecuteQuery
		(
			"UPDATE test_table " +
			"SET name = ? " +
			"WHERE amount = ?",
			WebFrameworkUtility.MakeSqlValues("empty", -1)
		);

		response.SetBody
		(
			new
			{
				data = table.ExecuteQuery<TableData>("SELECT * FROM test_table WHERE name = ?", WebFrameworkUtility.MakeSqlValues("empty"))
			}
		);
	}

	public override void DoDelete(HttpRequest request, HttpResponse response)
	{
		request.GetDatabase("test_database").GetTable("test_table").ExecuteQuery("DROP TABLE test_table");
	}
}

File: Tests/executors/CXX/database/src/CRUDExecutor.cpp
#include "CRUDExecutor.h"

#include <random>
#include <ctime>

#include <Utility/WebFrameworkUtility.hpp>

struct TableData
{
public:
	static TableData create(const framework::SqlResult::Row& row);

public:
	int64_t id;
	std::string name;
	int64_t amount;
};

void CRUDExecutor::init(const framework::utility::ExecutorSettings& settings)
{
	settings.getOrCreateDatabase("test_database").getOrCreateTable
	(
		"test_table",
		"CREATE TABLE IF NOT EXISTS test_table ("
		"id INTEGER PRIMARY KEY AUTOINCREMENT, "
		"name VARCHAR(255) NOT NULL, "
		"amount INTEGER NOT NULL)"
	);
}

void CRUDExecutor::doGet(framework::HttpRequest& request, framework::HttpResponse& response)
{
	framework::Database database = request.getDatabase("test_database");
	framework::Table table = database.getTable("test_table");
	std::vector<TableData> result = table.execute<TableData>("SELECT * FROM test_table WHERE name = ?", framework::utility::database::makeSqlValues("glue"));
	std::vector<framework::JsonObject> data;

	for (const TableData& value : result)
	{
		framework::JsonObject object;

		object["id"] = value.id;
		object["name"] = value.name;
		object["amount"] = value.amount;

		data.push_back(std::move(object));
	}

	response.setBody(framework::JsonBuilder().append("data", std::move(data)));
}

void CRUDExecutor::doPut(framework::HttpRequest& request, framework::HttpResponse& response)
{
	framework::Database database = request.getDatabase("test_database");
	framework::Table table = database.getTable("test_table");
	std::mt19937_64 random(std::time(nullptr));

	table.execute
	(
		"INSERT INTO test_table (name, amount) VALUES(?, ?)",
		framework::utility::database::makeSqlValues("glue", -1)
	);

	for (size_t i = 0; i < 10; i++)
	{
		table.execute
		(
			"INSERT INTO test_table (name, amount) VALUES(?, ?)",
			framework::utility::database::makeSqlValues("glue", random() % 200)
		);
	}
}

void CRUDExecutor::doPatch(framework::HttpRequest& request, framework::HttpResponse& response)
{
	framework::Database database = request.getDatabase("test_database");
	framework::Table table = database.getTable("test_table");

	table.execute
	(
		"UPDATE test_table "
		"SET name = ? "
		"WHERE amount = ?",
		framework::utility::database::makeSqlValues("empty", -1)
	);
	std::vector<TableData> result = table.execute<TableData>
	(
		"SELECT * FROM test_table WHERE name = ?",
		framework::utility::database::makeSqlValues("empty")
	);

	std::vector<framework::JsonObject> data;

	for (const TableData& value : result)
	{
		framework::JsonObject object;

		object["id"] = value.id;
		object["name"] = value.name;
		object["amount"] = value.amount;

		data.push_back(std::move(object));
	}

	response.setBody(framework::JsonBuilder().append("data", std::move(data)));
}

void CRUDExecutor::doDelete(framework::HttpRequest& request, framework::HttpResponse& response)
{
	request.getDatabase("test_database").getTable("test_table").execute("DROP TABLE test_table", {});
}

DEFINE_EXECUTOR(CRUDExecutor)

TableData TableData::create(const framework::SqlResult::Row& row)
{
	TableData result =
	{
		.id = row.at("id").get<int64_t>(),
		.name = row.at("name").get<std::string>(),
		.amount = row.at("amount").get<int64_t>()
	};

	return result;
}
